libs/datasets/dataset_orchestrator.py

```diff
@@ -0,0 +1,318 @@
+import json
+import datetime
+import structlog
+import pandas as pd
+from dataclasses import dataclass
+from typing import Dict, Iterable, List, Optional, Mapping
+
+from libs import pipeline
+from libs.parallel_utils import parallel_map
+from libs.datasets.sources import zeros_filter
+from libs.us_state_abbrev import US_STATE_ABBREV, ABBREV_US_UNKNOWN_COUNTY_FIPS
+from libs.datasets.timeseries import MultiRegionDataset
+from datapublic.common_fields import CommonFields, PdFields, FieldName
+from libs.datasets.dataset_utils import (
+    REGION_OVERRIDES_JSON,
+    CUMULATIVE_FIELDS_TO_FILTER,
+    DEFAULT_REPORTING_RATIO,
+)
+from libs.datasets import (
+    custom_patches,
+    weekly_hospitalizations,
+    nytimes_anomalies,
+    custom_aggregations,
+    statistical_areas,
+    timeseries,
+    outlier_detection,
+    dataset_utils,
+    combined_datasets,
+    new_cases_and_deaths,
+    vaccine_backfills,
+    tail_filter,
+    custom_patches,
+    weekly_hospitalizations,
+    manual_filter,
+    dataset_utils,
+    timeseries,
+    combined_datasets,
+)
+
+TailFilter = tail_filter.TailFilter
+_log = structlog.get_logger()
+
+# Putting this in dataset_utils caused a circular import with libs.pipeline
+KNOWN_LOCATION_ID_WITHOUT_POPULATION = [
+    # Territories other than PR
+    "iso1:us#iso2:us-vi",
+    "iso1:us#iso2:us-as",
+    "iso1:us#iso2:us-gu",
+    # Subregion of AS
+    "iso1:us#iso2:us-vi#fips:78030",
+    "iso1:us#iso2:us-vi#fips:78020",
+    "iso1:us#iso2:us-vi#fips:78010",
+    # Retired FIPS
+    "iso1:us#iso2:us-sd#fips:46113",
+    "iso1:us#iso2:us-va#fips:51515",
+    # All the unknown county FIPS
+    *[pipeline.fips_to_location_id(f) for f in ABBREV_US_UNKNOWN_COUNTY_FIPS.values()],
+]
+
+
+@dataclass
+class OneStateDataset:
+    """Wrapper class containing a MultiRegionDataset with data for a single state and its subregions. 
+    
+    As consistent with the Region class, territories are considered states (e.g. GU, PR)
+    """
+
+    multiregion_dataset: MultiRegionDataset
+    state: pipeline.Region
+
+    @classmethod
+    def from_mrds(cls, dataset: MultiRegionDataset) -> "OneStateDataset":
+        states = [pipeline.Region.from_location_id(loc).state for loc in dataset.location_ids]
+        if len(set(states)) != 1:
+            raise ValueError(
+                "Can only create a OneStateDataset from datasets with data "
+                f"for a single state. Data for {len(set(states))} found."
+            )
+        return OneStateDataset(
+            multiregion_dataset=dataset, state=pipeline.Region.from_state(states[0])
+        )
+
+
+@dataclass
+class MultiRegionOrchestrator:
+    """Class to orchestrate dataset updates through state-focused parallelization."""
+
+    regions: Iterable[OneStateDataset]
```

Comment:
nit: Given we have a "Region" class used pervasively through the codebase, calling this "regions" is a bit confusing.  Could we call it "state_datasets" or something?